tests/utils_tests/test_http.py

```diff
@@ -291,6 +292,44 @@ def test_secure_param_non_https_urls(self):
                     False,
                 )
 
+    # TODO: Remove when dropping support for PY38.
+    def test_invalid_bracketed_hosts(self):
+        # Port of urllib.parse.urlsplit() tests from Python.
+        tests = [
+            "Scheme://user@[192.0.2.146]/Path?Query",
+            "Scheme://user@[important.com:8000]/Path?Query",
+            "Scheme://user@[v123r.IP]/Path?Query",
+            "Scheme://user@[v12ae]/Path?Query",
+            "Scheme://user@[v.IP]/Path?Query",
+            "Scheme://user@[v123.]/Path?Query",
+            "Scheme://user@[v]/Path?Query",
+            "Scheme://user@[0439:23af::2309::fae7:1234]/Path?Query",
+            "Scheme://user@[0439:23af:2309::fae7:1234:2342:438e:192.0.2.146]/"
+            "Path?Query",
+            "Scheme://user@]v6a.ip[/Path",
+        ]
+        for invalid_url in tests:
+            with self.subTest(invalid_url=invalid_url):
+                self.assertRaises(ValueError, _urlsplit, invalid_url)
+
+    # TODO: Remove when dropping support for PY38.
+    def test_splitting_bracketed_hosts(self):
+        # Port of urllib.parse.urlsplit() tests from Python.
+        p1 = _urlsplit("scheme://user@[v6a.ip]/path?query")
+        self.assertEqual(p1.hostname, "v6a.ip")
+        self.assertEqual(p1.username, "user")
+        self.assertEqual(p1.path, "/path")
+        p2 = _urlsplit("scheme://user@[0439:23af:2309::fae7%test]/path?query")
+        self.assertEqual(p2.hostname, "0439:23af:2309::fae7%test")
+        self.assertEqual(p2.username, "user")
+        self.assertEqual(p2.path, "/path")
+        p3 = _urlsplit(
+            "scheme://user@[0439:23af:2309::fae7:1234:192.0.2.146%test]/path?query"
+        )
+        self.assertEqual(p3.hostname, "0439:23af:2309::fae7:1234:192.0.2.146%test")
+        self.assertEqual(p3.username, "user")
+        self.assertEqual(p3.path, "/path")
```

Comment:
It seems that `ipaddress.ip_address()` throws `ValueError` for `p2` and `p3` cases on Python 3.8. I think we can remove those cases because `test_invalid_bracketed_hosts()` is crucial for this backport.

Reply:
I _think_ this might be because 3.8 doesn't support scope id (https://github.com/python/cpython/commit/21da76d1f1b527d62b2e9ef79dd9aa514d996341) and p2 and p3 might use a scope id (see also this 3.6 backport https://github.com/openSUSE-Python/cpython/commit/4f2496bac3e516feddfab0da4dedaf61290aa37a)
So I agree that we could remove `p2` and `p3` (but I'm not confident I know what I'm talking about :+1:)

Reply:
Checked this morning and confirmed it was scope id stuff, adding the `p2` and `p3` case without the scope ids :+1: (so removing the `%test` suffixes)